JavaScript specials (#52)
javascript-tutorial/de.javascript.info: 1-js/02-first-steps/17-javascript-specials/article.md

[`confirm(question)`](mdn:api/Window/confirm)
: Stell eine `Frage` und schlag die Auswahlt zwischen OK und Abbrechen vor. Die Auswahl wird als `true/false` zurückgegeben.

[`alert(message)`](mdn:api/Window/alert)
: Ausgabe einer `Nachricht`.

Alle diese Funktionen sind *modal*, sie pausieren das Ausführen des Codes und verhindern, dass die Benutzer mit der Seite interagieren bis sie antworten.

Zum Beispiel:

```js run
let userName = prompt("Dein Name?", "Alice");
let isTeaWanted = confirm("Möchtest du etwas Tee?");

alert( "Besucher: " + userName ); // Alice
alert( "Tee gewünscht: " + isTeaWanted ); // true
```

Mehr in: <info:alert-prompt-confirm>.

## Operatoren

JavaScript unterstützt folgende Operatoren:

Arithmetisch
: Gewöhnlich: `* + - /`, ebenfalls `%` für den Rest und `**` für die Potenz einer Zahl.

    Das binäre Plus `+` setzt Zeichenketten zusammen. Und wenn einer der Operanden eine Zeichenkette ist, wird der andere auch in eine Zeichenkette konvertiert:

    ```js run
    alert( '1' + 2 ); // '12', string
    alert( 1 + '2' ); // '12', string
    ```

Zuweisungen
: Es gibt eine einfache Zuweisung: `a = b` und kombinierte `a *= 2`.

Bitweise

> : Bitweise Operatoren arbeiten mit 32-bit Ganzzaheln am niedrigsten bit-level: siehe die [docs](mdn:/JavaScript/Reference/Operators/Bitwise_Operators) wenn sie erforderlich sind.

Reviewer: ```suggestion
: Bitweise Operatoren arbeiten mit 32-bit Ganzzahlen am niedrigsten bit-level: siehe die [docs](mdn:/JavaScript/Reference/Operators/Bitwise_Operators), wenn sie erforderlich sind.
```

Bedingt
: Der einzige Operator mit drei Parametern: `cond ? resultA : resultB`. Wenn `cond` wahr ist, gib `resultA` zurück, andernfalls `resultB`.

Logische Operatoren

> : Logisches UND `&&` und ODER `||` führen Kurzzschlussauswertungen aus und dann geben sie den Wert zurück wo sie angehalten sind (nicht unbedingt `true`/`false`). Logisches NICHT `!` Konvertiert den Operand in einen bool'schen Typ und gibt den umgekehrten Wert zurück.

Reviewer: ```suggestion
: Logisch UND `&&` und ODER `||` führen eine Kurzzschluss-Auswertung durch und liefern dann den Wert zurück (nicht zwingend `true`/`false`), wo angehalten wurde. Logisch NICHT `!` konvertiert den Operand in einen boolschen Wert und gibt den invertierten Wert zurück.
```

Vergleiche
: Gleichheitsprüfung `==` für Werte unterschiedlichen Typs werden diese in eine Zahl umgewandelt (außer `null` und `undefined` die sich selber gleich sind aber zu nichts anderem), sodass diese gleich sind:

    ```js run
    alert( 0 == false ); // true
    alert( 0 == '' ); // true
    ```

    Andere Vergleiche konvertieren ebenfalls in eine Zahl.

>     Der strikte Gleichheitsoperator `===` führt diese Konversion nicht durch: unterschiedliche Typen bedeuten immer unterschieldiche Werte dafür.

Reviewer: ```suggestion
    Der strikte Gleichheitsoperator `===` führt diese Konvertierung nicht durch: verschiedene Typen bedeuten für ihn immer verschiedene Werte.
```

>     Werte `null` und `undefined` sind besonders: sie sind gleich `==` zueinander und nicht gleich zu etwas anderem.

Reviewer: ```suggestion
    Die Werte `null` und `undefined` sind etwas Besonderes: sie entsprechen `==` zueinander und sind mit nichts anderem gleichzusetzen.
```

>     Größer/kleiner-Vergleiche vergleichen Zeichenketten zeichenweise, andere Typen werden in eine Zahl umgewandelt.

Reviewer: ```suggestion
    Größer/kleiner-Vergleiche prüfen Zeichenketten zeichenweise, andere Typen werden in eine Zahl umgewandelt.
```

    }
    ```

> - Die Variable die in der `for(let...)`-Schleife deklariert wurde, ist nur in der Schleife sichtbar. Aber wir können `let` auch weglassen und eine vorhandene Variable wiederverwenden.

Reviewer: ```suggestion
- Die in der `for(let...)`-Schleife deklarierte Variable ist nur innerhalb der Schleife sichtbar. Wir können aber auch `let` weglassen und eine vorhandene Variable wiederverwenden.
```

> - Direktiven `break/continue` ermöglichen es die ganze Schleife/aktuelle Iteration zu verlassen. Benutze Labels um verschachtelte Schleifen abzubrechen.

Reviewer: ```suggestion
- Direktiven `break/continue` ermöglichen es die ganze Schleife/aktuelle Iteration zu verlassen. Benutze Labels, um verschachtelte Schleifen abzubrechen.
```

Details in: <info:while-for>.

> Später lernen wir weitere Typen von Schleifen kennen, um Objekte zu behandeln.

Reviewer: ```suggestion
Später werden wir weitere Arten von Schleifen untersuchen, um mit Objekten zu arbeiten.
```

## Das "switch" Konstrukt

Das "switch" Konstrukt kann mehrere `if`-Abfragen ersetzen. Es verwendet `===` (strikte Gleichheit) für Vergleiche.

Zum Beispiel:

```js run
let age = prompt('Dein Alter?', 18);

switch (age) {
  case 18:
    alert("Wird nicht funktionieren"); // Das Ergebnis der Eingabeaufforderung ist ein String, keine Zahl

  case "18":
    alert("Das funktioniert!");
    break;

  default:
    alert("Ein beliegibert Wert ungleich eines der oberen");
}
```

Details in: <info:switch>.

## Funktionen

> Wir haben drei Wege behandelt um Funktionen in JavaScript anzulegen:

Reviewer: ```suggestion
Wir betrachteten drei Möglichkeiten zum Erstellen einer Funktion in JavaScript:
```

    };
    ```

3. Pfeilfunktionen:

    ```js
    // Ausdruck auf der rechten Seite
    let sum = (a, b) => a + b;

    // oder mehrzeilige Syntax mit { ... }, benötigt hier return:
    let sum = (a, b) => {
      // ...
      return a + b;
    }

    // ohne Argumente
    let sayHi = () => alert("Hallo");

    // mit einem einzigen Argument
    let double = n => n * 2;
    ```

> - Funktionen können lokale Variablen haben: die innerhalb des Rumpfes deklariert wurden. Solche Variablen sind nur innerhalb der Funktion sichtbar.

Reviewer: ```suggestion
- Funktionen können lokale Variablen haben: diejenigen, die innerhalb ihres Körpers deklariert sind. Solche Variablen sind nur innerhalb der Funktion sichtbar.
```

- Parameter können Standardwerte haben: `function sum(a = 1, b = 2) {...}`.
- Funktionen geben immer etwas zurück. Wenn es keine `return` Anweisung gibt, dann ist das Ergebnis `undefined`.

Details: siehe <info:function-basics>, <info:arrow-functions-basics>.

## Da kommt noch mehr

> Das war eine kurze Liste von JavaScript-Features. Ab hier haben wir nur die Grundlagen gelernt. Weiter im Tutorial wirst du mehr Besonderheiten und fortgeschrittene Features von JavaScript finden.

Reviewer: ```suggestion
Das war eine kurze Zusammenfassung von JavaScript-Funktionen. Bis jetzt haben wir nur die Grundlagen kennengelernt. Im weiteren Verlauf des Tutorials findest du weitere Besonderheiten und fortgeschrittene Funktionen von JavaScript.
```
